pkg/controller/database/mysqlserver/managed.go

```diff
@@ -0,0 +1,212 @@
+	errCreateMysqlServer    = "cannot create MysqlServer"
+	errGetMysqlServer       = "cannot get MysqlServer"
+	errDeleteMysqlServer    = "cannot delete MysqlServer"
+	errCheckMysqlServerName = "cannot check MysqlServer name availability"
+)
+
+// Controller is responsible for adding the MysqlServer controller and its
+// corresponding reconciler to the manager with any runtime configuration.
+type Controller struct{}
+
+// SetupWithManager creates a new MysqlServer Controller and adds it to the
+// Manager with default RBAC. The Manager will set fields on the Controller and
+// start it when the Manager is Started.
+func (c *Controller) SetupWithManager(mgr ctrl.Manager) error {
+	r := resource.NewManagedReconciler(mgr,
+		resource.ManagedKind(v1alpha2.MysqlServerGroupVersionKind),
+		resource.WithExternalConnecter(&connecter{client: mgr.GetClient(), newClientFn: newClient}))
+
+	name := strings.ToLower(fmt.Sprintf("%s.%s", v1alpha2.MysqlServerKind, v1alpha2.Group))
+
+	return ctrl.NewControllerManagedBy(mgr).
+		Named(name).
+		For(&v1alpha2.MysqlServer{}).
+		Complete(r)
+}
+
+func newClient(credentials []byte) (azure.MySQLServerAPI, error) {
+	ac, err := azure.NewClient(credentials)
+	if err != nil {
+		return nil, errors.Wrap(err, "cannot create Azure client")
+	}
+	mc, err := azure.NewMySQLServerClient(ac)
+	return mc, errors.Wrap(err, "cannot create Azure MySQL client")
+}
+
+type connecter struct {
+	client      client.Client
+	newClientFn func(credentials []byte) (azure.MySQLServerAPI, error)
+}
+
+func (c *connecter) Connect(ctx context.Context, mg resource.Managed) (resource.ExternalClient, error) {
+	v, ok := mg.(*v1alpha2.MysqlServer)
+	if !ok {
+		return nil, errors.New(errNotMysqlServer)
+	}
+
+	p := &azurev1alpha2.Provider{}
+	if err := c.client.Get(ctx, meta.NamespacedNameOf(v.Spec.ProviderReference), p); err != nil {
+		return nil, errors.Wrap(err, errGetProvider)
+	}
+
+	s := &corev1.Secret{}
+	n := types.NamespacedName{Namespace: p.Spec.Secret.Namespace, Name: p.Spec.Secret.Name}
+	if err := c.client.Get(ctx, n, s); err != nil {
+		return nil, errors.Wrap(err, errGetProviderSecret)
+	}
+	client, err := c.newClientFn(s.Data[p.Spec.Secret.Key])
+	return &external{client: client, newPasswordFn: util.GeneratePassword}, errors.Wrap(err, errNewClient)
+}
+
+type external struct {
+	client        azure.MySQLServerAPI
+	newPasswordFn func(len int) (password string, err error)
+}
+
+func (e *external) Observe(ctx context.Context, mg resource.Managed) (resource.ExternalObservation, error) {
+	s, ok := mg.(*v1alpha2.MysqlServer)
+	if !ok {
+		return resource.ExternalObservation{}, errors.New(errNotMysqlServer)
+	}
+
+	external, err := e.client.GetServer(ctx, s)
+	if azure.IsNotFound(err) {
+		// Azure SQL servers don't exist according to the Azure API until their
+		// create operation has completed, and Azure will happily let you submit
+		// several subsequent create operations for the same server. Our create
```

Comment:
Wow, I did not know this. Thank you for identifying this! I do not love this kind of API behavior by the Azure SDK :/

Reply:
Yeah it's not great. Without this check we'd repeatedly try to create the server. The first create would win, and the other 3-4 would fail, but we'd generate a new random password each time and end up persisting a password that did not actually work.